pkg/security/admission/admission.go

```diff
@@ -0,0 +1,390 @@
+			if err != nil {
+				return nil, fmt.Errorf("unable to find pre-allocated uid annotation for namespace %s while trying to configure SCC %s: %v", namespace.Name, constraint.Name, err)
+			}
+			constraint.RunAsUser.UIDRangeMin = min
+			constraint.RunAsUser.UIDRangeMax = max
+		}
+
+		if requiresPreAllocatedSELinuxLevel(constraint) {
+			namespace, err = c.getNamespace(ns, namespace)
+			if err != nil {
+				return nil, err
+			}
+
+			// if we don't have options create them here so we can add the preallocated settings
+			if constraint.SELinuxContext.SELinuxOptions == nil {
+				constraint.SELinuxContext.SELinuxOptions = &kapi.SELinuxOptions{}
+			}
+
+			level, err := getPreallocatedLevel(namespace)
+			if err != nil {
+				return nil, fmt.Errorf("unable to find pre-allocated mcs annotation for namespace %s while trying to configure SCC %s: %v", namespace.Name, constraint.Name, err)
+			}
+			constraint.SELinuxContext.SELinuxOptions.Level = level
+		}
+
+		provider, err := scc.NewSimpleProvider(constraint)
+		if err != nil {
+			return nil, fmt.Errorf("error creating provider for SCC %s in namespace %s: %v", constraint.Name, ns, err)
+		}
+		providers = append(providers, provider)
+	}
+	return providers, nil
+}
+
+// getSecurityContextConstraintsForServiceAccount retrieves SCCs for the service account if a
+// service account is specified on the pod.
+func (c *constraint) getSecurityContextConstraintsForServiceAccount(ns string, pod *kapi.Pod) ([]*kapi.SecurityContextConstraints, error) {
+	// if there is a service account get the constraints for the service account
+	if len(pod.Spec.ServiceAccount) > 0 {
+		userInfo := serviceaccount.UserInfo(ns, pod.Spec.ServiceAccount, "")
+		glog.V(3).Infof("getting security context constraints for pod %s with service account info %v", pod.Name, userInfo)
+		return getMatchingSecurityContextConstraints(c.store, userInfo)
+	}
+	return []*kapi.SecurityContextConstraints{}, nil
+}
+
+// getNamespace retrieves a namespace only if ns is nil.
+func (c *constraint) getNamespace(name string, ns *kapi.Namespace) (*kapi.Namespace, error) {
+	if ns != nil {
+		return ns, nil
+	}
+	return c.client.Namespaces().Get(name)
+}
+
+// getMatchingSecurityContextConstraints returns constraints from the store that match the group,
+// uid, or user of the service account.
+func getMatchingSecurityContextConstraints(store cache.Store, userInfo user.Info) ([]*kapi.SecurityContextConstraints, error) {
+	matchedConstraints := make([]*kapi.SecurityContextConstraints, 0)
+
+	for _, c := range store.List() {
+		constraint, ok := c.(*kapi.SecurityContextConstraints)
+		if !ok {
+			return nil, errors.NewInternalError(fmt.Errorf("error converting object from store to a security context constraint: %v", c))
+		}
+		if constraintAppliesTo(constraint, userInfo) {
+			// don't pass the pointer, pass a copy so that if pre-allocated values are set they
+			// don't affect the store.
+			copy, err := kapi.Scheme.Copy(constraint)
```

Comment:
not a fan of copying here... if an SCC applies to both user and service account, we copy it twice. if ten SCC's apply and the first one validates, we copied the other nine unnecessarily.